manager/allocator/network/port/port_test.go

```diff
@@ -0,0 +1,955 @@
+				spec := endpoint.Spec.Copy()
+				spec.Ports[0].TargetPort = uint32(999)
+				spec.Ports[0].Name = "something else"
+				p, _ := pa.Allocate(endpoint, spec)
+
+				Expect(p.Ports()[0].TargetPort).To(Equal(uint32(999)))
+				Expect(p.Ports()[0].Name).To(Equal("something else"))
+			})
+		})
+		Context("to change published port", func() {
+			var (
+				endpointCopy *api.Endpoint
+				p            Proposal
+				err          error
+			)
+			BeforeEach(func() {
+				endpointCopy = endpoint.Copy()
+				spec := endpoint.Spec.Copy()
+				spec.Ports[0].PublishedPort = uint32(999)
+				p, err = pa.Allocate(endpointCopy, spec)
+			})
+			It("should allocate the new port", func() {
+				Expect(err).ToNot(HaveOccurred())
+				Expect(p).ToNot(BeNil())
+				Expect(p.IsNoop()).To(BeFalse())
+				Expect(p.Ports()[0].PublishedPort).To(Equal(uint32(999)))
+
+				p.Commit()
+
+				// try allocating the port we just allocated
+				_, err := pa.Allocate(&api.Endpoint{}, &api.EndpointSpec{
+					Ports: []*api.PortConfig{
+						p.Ports()[0],
+					},
+				})
+				Expect(err).To(HaveOccurred())
+				Expect(IsErrPortInUse(err)).To(BeTrue())
+			})
+			It("should free the old port", func() {
+				p.Commit()
+
+				q, err := pa.Allocate(&api.Endpoint{}, &api.EndpointSpec{
+					Ports: []*api.PortConfig{
+						{
+							Protocol:      api.ProtocolTCP,
+							TargetPort:    80,
+							PublishedPort: endpoint.Ports[0].PublishedPort,
+						},
+					},
+				})
+				Expect(err).ToNot(HaveOccurred())
+				Expect(q.IsNoop()).To(BeFalse())
+			})
+			It("should not free any other ports", func() {
+				p.Commit()
+
+				_, err := pa.Allocate(&api.Endpoint{}, &api.EndpointSpec{
+					Ports: p.Ports()[1:],
+				})
+				Expect(err).To(HaveOccurred())
+				Expect(IsErrPortInUse(err)).To(BeTrue())
+			})
+		})
+		Context("to change protocol", func() {
+			var (
+				endpointCopy *api.Endpoint
+				p            Proposal
+				err          error
+			)
+			BeforeEach(func() {
+				endpointCopy = endpoint.Copy()
+				spec := endpoint.Spec.Copy()
+				spec.Ports[0].Protocol = api.ProtocolSCTP
+				p, err = pa.Allocate(endpointCopy, spec)
+			})
+			It("should allocate the new port", func() {
+				Expect(err).ToNot(HaveOccurred())
+				Expect(p.IsNoop()).ToNot(BeTrue())
+
+				e := endpoint.Copy()
+				e.Ports[0].Protocol = api.ProtocolSCTP
+				Expect(p.Ports()).To(ConsistOf(e.Ports))
+
+				p.Commit()
+				_, err := pa.Allocate(&api.Endpoint{}, &api.EndpointSpec{
+					Ports: p.Ports()[0:1],
+				})
+				Expect(err).To(HaveOccurred())
+				Expect(IsErrPortInUse(err)).To(BeTrue())
+			})
+			It("should free the old port", func() {
+				p.Commit()
+				_, err := pa.Allocate(&api.Endpoint{}, &api.EndpointSpec{
+					Ports: endpoint.Ports[0:1],
+				})
+				Expect(err).ToNot(HaveOccurred())
+			})
+		})
+		// Pending, we need to test this but I'm not sure what the appropriate
+		// behavior should be
+		PContext("to change publish mode", func() {
```

Comment:
Are we still deciding what the appropriate behavior should be?  It seems like nothing gets deallocated or allocated (so if it previously had an allocated port, changing the publish mode to host would not deallocate the previous port).

Reply:
Must've missed this when I was going through, I will add these test cases, thanks!